File: backend/controllers/payment_controller.py
# Conceptual Implementation for Orange Money Payment Integration
# This is a highly conceptual outline and depends on the actual Orange Money API.

# --- Assumed Orange Money API Functionalities ---
# 1. Initiate Payment:
#    - Endpoint: (e.g., POST https://api.orange-money.com/v1/initiate_payment)
#    - Request: { "amount": 1000, "currency": "XOF", "customer_phone": "22177xxxxxxx", "order_id": "invoice_id_123", "callback_url": "https://yourdomain.com/api/payments/orangemoney/callback", "merchant_key": "YOUR_MERCHANT_KEY" }
#    - Response: { "transaction_id": "OM_TXN_12345", "status": "PENDING", "payment_url": "https://orange-money.com/pay?txn=OM_TXN_12345" } or USSD instructions.
# 2. Check Payment Status:
#    - Endpoint: (e.g., GET https://api.orange-money.com/v1/payment_status?transaction_id=OM_TXN_12345)
#    - Response: { "transaction_id": "OM_TXN_12345", "status": "SUCCESSFUL" | "FAILED" | "PENDING", ... }

# --- Security and Configuration Notes ---
# - API Keys/Secrets: Orange Money API credentials (merchant key, API key/secret) must be stored securely, typically as environment variables.
# - Callback URLs: Must be publicly accessible HTTPS endpoints. Whitelisting by Orange Money might be required.
# - Callback Verification: Orange Money callbacks must be verified (e.g., checking a signature passed in headers, verifying the source IP) to prevent spoofing.
# - Idempotency: Callback handlers should be idempotent (i.e., processing the same callback multiple times should not cause issues).
# - Logging: Detailed logging of payment initiation, callback reception, and status updates is crucial for debugging.

# --- API Endpoints for Our Application ---

# 1. Initiate Orange Money Payment for an Invoice
# Endpoint: POST /api/invoices/{invoice_id}/pay/orangemoney
# Description: Initiates a payment for a specific invoice using Orange Money.
# Request Payload (could be empty if all info is derived from invoice_id and user session, or could include phone number):
# {
#   "customer_phone": "22177xxxxxxx" // Phone number to charge for Orange Money
# }
# Response Payload (Success - 200 OK):
# {
#   "payment_url": "https://orange-money.com/pay?txn=OM_TXN_12345", // If Orange Money provides a redirect URL
#   "transaction_id": "OM_TXN_12345",
#   "message": "Payment initiated. Please follow the instructions to complete.",
#   "payment_status": "pending" // Current status of the invoice's payment attempt
# }
# Response Payload (Error - e.g., 400 Bad Request, 404 Invoice Not Found, 500 API Error):
# { "error": "Detailed error message." }
# Conceptual Steps:
# 1. Authenticate and authorize the user.
# 2. Retrieve the invoice by `invoice_id`. Ensure it's unpaid and valid.
# 3. Get customer's phone number (from request or user profile).
# 4. Construct payload for Orange Money's "Initiate Payment" API (amount, currency, phone, our callback URL, our invoice_id as order_id).
# 5. Call Orange Money API.
# 6. If successful:
#    a. Store the returned `transaction_id` on the invoice (e.g., `invoice.payment_transaction_id = "OM_TXN_12345"`).
#    b. Update invoice's payment status (e.g., `invoice.payment_status = 'pending_orangemoney'`).
#    c. Save invoice changes to the database.
#    d. Return the `payment_url` (if any) or instructions to the frontend.
# 7. If failed, return an appropriate error response.

def initiate_orange_money_payment_handler(invoice_id, request_data, user):
    # Conceptual:
    # invoice = db.get_invoice(invoice_id)
    # if not invoice or invoice.user_id != user.id: return {"error": "Invoice not found or unauthorized"}, 404
    # if invoice.status == 'paid': return {"error": "Invoice already paid"}, 400
    #
    # customer_phone = request_data.get("customer_phone") # Or from user.profile.phone
    # if not customer_phone: return {"error": "Customer phone number required"}, 400
    #
    # orange_money_payload = {
    #     "amount": invoice.total_amount,
    #     "currency": invoice.currency, # Ensure this is XOF or other OM supported currency
    #     "customer_phone": customer_phone,
    #     "order_id": invoice.id, # Our internal ID for reconciliation
    #     "callback_url": "https://yourdomain.com/api/payments/orangemoney/callback",
    #     "merchant_key": os.environ.get("ORANGE_MONEY_MERCHANT_KEY")
    # }
    #
    # try:
    #   om_response = requests.post("https://api.orange-money.com/v1/initiate_payment", json=orange_money_payload)
    #   om_response.raise_for_status() # Raise an exception for HTTP error codes
    #   om_data = om_response.json()
    # except requests.RequestException as e:
    #   # Log error
    #   return {"error": "Failed to initiate payment with Orange Money."}, 500
    #
    # invoice.payment_method = 'orangemoney'
    # invoice.payment_transaction_id = om_data.get("transaction_id")
    # invoice.payment_status = 'pending' # Or a more specific 'pending_orangemoney'
    # invoice.status = 'payment_pending' # Update master invoice status
    # db.save_invoice(invoice)
    #
    # return {
    #   "payment_url": om_data.get("payment_url"), # Or USSD code/instructions
    #   "transaction_id": om_data.get("transaction_id"),
    #   "message": "Payment initiated. Please follow the instructions.",
    #   "payment_status": invoice.payment_status
    # }, 200
    pass


# 2. Orange Money Callback/Webhook Handler
# Endpoint: POST /api/payments/orangemoney/callback
# Description: Receives notifications from Orange Money about payment status updates.
# Request Payload (from Orange Money - example, actual payload varies):
# {
#   "transaction_id": "OM_TXN_12345",
#   "order_id": "invoice_id_123", // Our internal ID
#   "status": "SUCCESSFUL", // or "FAILED", "EXPIRED"
#   "amount": 1000,
#   "currency": "XOF",
#   "timestamp": "2024-07-31T12:30:00Z",
#   "signature": "orange_money_generated_signature" // For verifying authenticity
# }
# Response Payload (to Orange Money - typically 200 OK for acknowledgment):
# {}
# Conceptual Steps:
# 1. Verify the callback's authenticity (e.g., check signature, source IP). This is CRITICAL.
#    - `is_valid_signature = verify_orange_money_signature(request.headers, request.body, os.environ.get("ORANGE_MONEY_SECRET"))`
#    - If not valid, return 400 Bad Request or 403 Forbidden and do not process.
# 2. Extract `transaction_id` and `order_id` (our invoice ID) and payment `status` from the payload.
# 3. Retrieve the invoice using `order_id`.
# 4. If invoice found and `invoice.payment_transaction_id` matches the callback's `transaction_id`:
#    a. If Orange Money status is "SUCCESSFUL":
#       i.   Update `invoice.payment_status = 'successful'`.
#       ii.  Update `invoice.status = 'paid'`.
#       iii. Record payment details (amount, time, etc.).
#       iv.  Trigger post-payment actions (e.g., send receipt, notify user).
#    b. If Orange Money status is "FAILED" or "EXPIRED":
#       i.   Update `invoice.payment_status = 'failed'` (or 'expired').
#       ii.  Optionally, revert `invoice.status` to 'sent' or 'overdue' or a specific 'payment_failed' status.
#    c. Save invoice changes to the database.
# 5. Respond to Orange Money with a 200 OK to acknowledge receipt. If not acknowledged, Orange Money might retry the callback.

def orange_money_callback_handler(request_data, request_headers):
    # Conceptual:
    # if not verify_orange_money_signature(request_headers, request_data):
    #    # Log security event
    #    return {"error": "Invalid signature"}, 403
    #
    # transaction_id = request_data.get("transaction_id")
    # order_id = request_data.get("order_id") # Our invoice ID
    # om_status = request_data.get("status")
    #
    # invoice = db.get_invoice_by_order_id(order_id) # Or by transaction_id if stored before
    # if not invoice:
    #    # Log: Received callback for unknown order/invoice
    #    return {"error": "Invoice not found for order_id"}, 404 # Or 200 OK if OM requires it for unknown orders
    #
    # if invoice.payment_transaction_id != transaction_id:
    #    # Log: Mismatched transaction ID
    #    return {"error": "Transaction ID mismatch"}, 400
    #
    # if om_status == "SUCCESSFUL":
    #    if invoice.status == 'paid': # Idempotency check
    #        # Log: Invoice already marked as paid
    #        return {}, 200
    #    invoice.payment_status = 'successful'
    #    invoice.status = 'paid'
    #    # invoice.paid_at = datetime.utcnow()
    #    # record_payment_event(invoice, request_data)
    #    # send_receipt_email(invoice)
    # elif om_status in ["FAILED", "EXPIRED"]:
    #    invoice.payment_status = om_status.lower() # 'failed' or 'expired'
    #    # invoice.status = 'payment_failed' # Or revert to 'sent' / 'overdue'
    # else:
    #    # Log: Unhandled Orange Money status
    #    invoice.payment_status = f"unknown_{om_status.lower()}"
    #
    # db.save_invoice(invoice)
    # return {}, 200 # Acknowledge receipt to Orange Money
    pass

# 3. (Optional) Get Invoice Payment Status
# Endpoint: GET /api/invoices/{invoice_id}/payment_status
# Description: Allows the frontend to poll for the payment status of an invoice.
# Response Payload (Success - 200 OK):
# {
#   "invoice_id": "inv_id_1",
#   "payment_status": "successful", # or 'pending', 'failed', 'paid' (from invoice.status)
#   "status": "paid"
# }
# Conceptual Steps:
# 1. Authenticate and authorize user.
# 2. Retrieve invoice.
# 3. Return relevant status fields (`payment_status`, `status`).

def get_invoice_payment_status_handler(invoice_id, user):
    # Conceptual:
    # invoice = db.get_invoice(invoice_id)
    # if not invoice or invoice.user_id != user.id: return {"error": "Invoice not found or unauthorized"}, 404
    #
    # return {
    #   "invoice_id": invoice.id,
    #   "payment_status": invoice.payment_status, # The specific status from the payment gateway attempt
    #   "invoice_status": invoice.status         # The overall status of the invoice
    # }, 200
    pass

File: backend/controllers/paymentController.js
// Placeholder for Payment Controller

// const paymentService = require('../services/paymentService');
// const invoiceService = require('../services/invoiceService'); // To get invoice details if needed
// const { validationResult } = require('express-validator'); // For request validation

const paymentController = {
  /**
   * Initiates a payment for a specific invoice.
   * POST /api/invoices/:invoiceId/pay
   */
  initiatePaymentForInvoice: async (req, res) => {
    // const errors = validationResult(req);
    // if (!errors.isEmpty()) {
    //   return res.status(400).json({ errors: errors.array() });
    // }

    try {
      const { invoiceId } = req.params;
      const { paymentMethod, customerMsisdn, currency, amount } = req.body; // paymentMethod could be 'orangemoney_senegal'
      const userId = req.user ? req.user.id : 'test_user_id'; // From auth middleware

      if (!paymentMethod || paymentMethod !== 'orangemoney_senegal') {
        return res.status(400).json({ message: 'Invalid or unsupported payment method specified.' });
      }

      // In a real app, amount and currency might be fetched from the invoice itself
      // by the service to prevent tampering, rather than taken directly from req.body for payment.
      // For now, we assume they are validated or passed correctly for the service to use/validate.
      if (!customerMsisdn || !amount || !currency) {
          return res.status(400).json({ message: 'Missing required fields: customerMsisdn, amount, or currency for payment.' });
      }


      // Call the payment service
      // const paymentInitiationResult = await paymentService.initiateOrangeMoneyPayment(
      //   invoiceId,
      //   parseFloat(amount), // Ensure amount is a number
      //   currency,
      //   customerMsisdn,
      //   userId
      // );

      // Simulating service call
      console.log(`PaymentController: User ${userId} initiating ${paymentMethod} payment for invoice ${invoiceId} with MSISDN ${customerMsisdn}, Amount: ${amount} ${currency}`);
      const paymentInitiationResult = { // Mock response from paymentService.initiateOrangeMoneyPayment
          message: 'Payment initiated with Orange Money (Simulated). Follow instructions.',
          transactionId: `OM_TXN_CTRL_${Date.now()}`,
          orderId: `INV-${invoiceId}-${Date.now()}`,
          // paymentUrl: `https://simulated-om.sn/pay?txn=${Date.now()}` // Example
      };

      res.status(200).json(paymentInitiationResult);

    } catch (error) {
      // console.error('Initiate payment error:', error);
      // res.status(500).json({ message: `Error initiating payment: ${error.message}` });
      res.status(500).json({ message: `Error initiating payment (placeholder): ${error.message}` });
    }
  },

  /**
   * Handles incoming notifications/callbacks from Orange Money.
   * POST /api/webhooks/orangemoney/senegal/notification
   */
  orangeMoneyNotificationWebhook: async (req, res) => {
    try {
      const callbackData = req.body;
      console.log('PaymentController: Received Orange Money webhook notification:', callbackData);

      // It's crucial that the service layer handles all business logic, including verification.
      // await paymentService.handleOrangeMoneyCallback(callbackData);

      // Simulate service call
      const { order_id, transaction_id, status: gatewayStatus } = callbackData;
      if (!order_id || !transaction_id || !gatewayStatus) {
          console.warn("PaymentController: Invalid callback data received from Orange Money.");
          return res.status(400).json({ message: "Invalid callback data."});
      }
      console.log(`PaymentController: Passing callback for order ${order_id} (txn: ${transaction_id}, status: ${gatewayStatus}) to service.`);


      // Respond quickly to the webhook provider
      res.status(200).json({ message: 'Webhook received successfully.' });

    } catch (error) {
      // console.error('Orange Money webhook error:', error);
      // Even on error, usually respond with 200 if possible, or a specific error code
      // if the gateway requires it for retry mechanisms.
      // res.status(500).json({ message: `Error processing Orange Money webhook: ${error.message}` });
      res.status(500).json({ message: `Error processing Orange Money webhook (placeholder): ${error.message}` });
    }
  },

  /**
   * Placeholder for checking payment status, might be called by frontend or a scheduled job.
   * GET /api/payments/orangemoney/status/:transactionId (Example route)
   */
  checkPaymentStatus: async (req, res) => {
      try {
          const { transactionId } = req.params;
          // const statusDetails = await paymentService.checkOrangeMoneyPaymentStatus(transactionId);

          // Simulate service call
          const statusDetails = { // Mock response from paymentService.checkOrangeMoneyPaymentStatus
              transaction_id: transactionId,
              status: 'SUCCESS',
              amount: 1500.00,
              currency: 'XOF',
              timestamp: new Date().toISOString(),
          };
          console.log(`PaymentController: Checked status for OM txn ${transactionId}. Status: ${statusDetails.status}`);

          res.status(200).json(statusDetails);
      } catch (error) {
          // console.error('Check payment status error:', error);
          // res.status(500).json({ message: `Error checking payment status: ${error.message}` });
          res.status(500).json({ message: `Error checking payment status (placeholder): ${error.message}` });
      }
  }
};

module.exports = paymentController;
